[UnifiedPDF] [iOS] Support text selection adjustment with selection handles

whsieh opened the pull request:
#### fc167fd5d50cfa0a26daf54b11a68055355d85d5
<pre>
[UnifiedPDF] [iOS] Support text selection adjustment with selection handles
<a href="https://bugs.webkit.org/show_bug.cgi?id=284582">https://bugs.webkit.org/show_bug.cgi?id=284582</a>
<a href="https://rdar.apple.com/141324936">rdar://141324936</a>

Reviewed by Abrar Rahman Protyasha.

Add support for range adjustment text interactions (i.e. dragging selection handles to select text)
when using unified PDF on iOS. See below for more details.

* Source/WebKit/Shared/ios/GestureTypes.h:

Add a couple new selection-related enum types to help clarify the parameter/return type in the new
`moveSelectionEndpoint` method (see below).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::moveSelectionEndpoint):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setSelectionRange):

Adopt the new `rootViewToPage` helper below.

(WebKit::UnifiedPDFPlugin::moveSelectionEndpoint):

Add a new helper method to handle selection changes triggered by the user moving one of the
selection endpoints with a loupe gesture on iOS. This handles selection flipping by detecting when
the location of the selection base (i.e. the stationary selection range endpoint) has moved, and
surfaces `SelectionWasFlipped::Yes` back to UIKit in order to allow the system text interaction UI
to swap the selection start and end handles on the fly.

(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage):
(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage const):

Add a couple of helper methods to compute the selection start/end points, relative to page
coordinates. Returns both a point, and a `PDFPage` that is the the point&apos;s coordinate space.

(WebKit::UnifiedPDFPlugin::rootViewToPage const):

Pull out some existing code to map a point from root view coordinates to a PDF page out into a
separate helper. This returns both the new mapped point, as well as the `PDFPage` that the point is
relative to.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::moveSelectionEndpoint):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateSelectionWithTouches):

Add a hook to defer to the PDF plugin, if present.

Canonical link: <a href="https://commits.webkit.org/287780@main">https://commits.webkit.org/287780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd5e278a254e7053104fd62a95accb6a753678c6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34718 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85309 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31766 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/321 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8102 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63086 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20875 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83851 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/149 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73527 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43388 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/75 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30223 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28219 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86741 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8009 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71382 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8186 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69363 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70621 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17594 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14648 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7971 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7810 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11329 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9616 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3900,19 +3901,100 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+    if (!basePageAndPoint)
+        return flipped;
+
+    auto extentPageAndPoint = selectionCaretPointInPage(extentEndpoint);
```
Is this not used anywhere?

whsieh replied:
🤦 Good catch, this should be removed.

Thank you for the review!

whsieh committed:
[UnifiedPDF] [iOS] Support text selection adjustment with selection handles
https://bugs.webkit.org/show_bug.cgi?id=284582
rdar://141324936

Reviewed by Abrar Rahman Protyasha.

Add support for range adjustment text interactions (i.e. dragging selection handles to select text)
when using unified PDF on iOS. See below for more details.

* Source/WebKit/Shared/ios/GestureTypes.h:

Add a couple new selection-related enum types to help clarify the parameter/return type in the new
`moveSelectionEndpoint` method (see below).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::moveSelectionEndpoint):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setSelectionRange):

Adopt the new `rootViewToPage` helper below.

(WebKit::UnifiedPDFPlugin::moveSelectionEndpoint):

Add a new helper method to handle selection changes triggered by the user moving one of the
selection endpoints with a loupe gesture on iOS. This handles selection flipping by detecting when
the location of the selection base (i.e. the stationary selection range endpoint) has moved, and
surfaces `SelectionWasFlipped::Yes` back to UIKit in order to allow the system text interaction UI
to swap the selection start and end handles on the fly.

(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage):
(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage const):

Add a couple of helper methods to compute the selection start/end points, relative to page
coordinates. Returns both a point, and a `PDFPage` that is the the point's coordinate space.

(WebKit::UnifiedPDFPlugin::rootViewToPage const):

Pull out some existing code to map a point from root view coordinates to a PDF page out into a
separate helper. This returns both the new mapped point, as well as the `PDFPage` that the point is
relative to.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::moveSelectionEndpoint):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateSelectionWithTouches):

Add a hook to defer to the PDF plugin, if present.

Canonical link: https://commits.webkit.org/287780@main
